lib/AST/GenerateAST.cpp

```diff
@@ -236,36 +236,68 @@ z3::expr GenerateAST::GetOrCreateEdgeCond(llvm::BasicBlock *from,
   return z_exprs[z_edges[{from, to}]];
 }
 
-z3::expr GenerateAST::GetOrCreateReachingCond(llvm::BasicBlock *block) {
+z3::expr GenerateAST::GetReachingCond(llvm::BasicBlock *block) {
   if (reaching_conds.find(block) == reaching_conds.end()) {
-    if (block->hasNPredecessorsOrMore(1)) {
-      auto cond{z_ctx->bool_val(false)};
-      // Gather reaching conditions from predecessors of the block
-      for (auto pred : llvm::predecessors(block)) {
-        auto has_pred_cond{reaching_conds.find(pred) != reaching_conds.end()};
-        auto edge_cond{GetOrCreateEdgeCond(pred, block)};
-        // Construct reaching condition from `pred` to `block` as
-        // `reach_cond[pred] && edge_cond(pred, block)` or one of
-        // the two if the other one is missing.
-        auto conj_cond{has_pred_cond
-                           ? (z_exprs[reaching_conds[pred]] && edge_cond)
-                           : edge_cond};
-        // Append `conj_cond` to reaching conditions of other
-        // predecessors via an `||`. Use `conj_cond` if there
-        // is no `cond` yet.
-        cond = cond || conj_cond;
-      }
+    return z_ctx->bool_val(false);
+  }
 
-      reaching_conds[block] = z_exprs.size();
-      z_exprs.push_back(cond.simplify());
-    } else {
-      auto cond{z_ctx->bool_val(true)};
+  return z_exprs[reaching_conds[block]];
+}
+
+void GenerateAST::CreateReachingCond(llvm::BasicBlock *block) {
+  auto Prove{[this](z3::expr expr) {
+    z3::goal goal(*z_ctx);
+    goal.add((!expr).simplify());
+    auto app{z3::tactic(*z_ctx, "sat").apply(goal)};
+    CHECK(app.size() == 1) << "Unexpected multiple goals in application!";
+    return app[0].is_decided_unsat();
+  }};
+
+  auto Simplify{[this, &Prove](z3::expr expr) {
+    if (Prove(expr)) {
+      return z_ctx->bool_val(true);
+    }
+
+    z3::tactic aig(*z_ctx, "aig");
+    z3::tactic simplify(*z_ctx, "simplify");
+    z3::tactic ctx_solver_simplify(*z_ctx, "ctx-solver-simplify");
```

Comment:
This is a pretty heavy duty simplifier :eyes: Have you tried `ctx-simplify`? Maybe mix `solve-eqs` in?

Reply:
`solve-eqs` seems to do the trick!